### ISSUE TITLE
Concurrent Versions of Reduction Operations

### ISSUE THREAD

**Issue by mtoohey31** [mtoohey31/iter#14]
It may be possible to create concurrent versions of many other reduction functions, like we've already done for `ForEachCon`:

- [x] `Consume` d5ce279edc8d8ff5f7c80981aa97c858cd7cd26a
- [x] `Collect` and `CollectInto` cdc8cb11390d8400d16d5a9dae671fe862dc86a8
- [x] `All` 0310170a2733583b020ff66c6ca42ce60c176f9c
- [x] `Any` 0310170a2733583b020ff66c6ca42ce60c176f9c
- [x] `Count` 32ad517d17092210794bb4cb46b7f111bdf0c6a1
- [ ] `Find` and variants (will require consuming the whole iterator)
- [ ] `Fold` and variants (`FoldAC` - associative and commutative, since we can parallelize the reduction when we have associativity and commutativity)
- [ ] `Last`
- [ ] `TryFold` and variants
- [ ] `TryForEach`
- [ ] `Reduce` (`AC` variant too, as with `Fold`)
- [ ] `Position`

Some of the above may not be possible, I haven't thought things through completely. Implementing some of these may also reveal thread safety issues with the implementations of the non-reductive functions, since they haven't been tested concurrently.

**Comment by mtoohey31** [comment]
As noted in 2b38251d8fdb48310af87592248a23d905d86953, I have decided that these features should be part of a separate library instead. This issue will be updated and closed once I create that library.

**Comment by mtoohey31** [comment]
I've decided that this isn't really a good idea. The main advantage of having concurrent iterators would be performance, but doing things with iterators won't be the most performance solution in most cases anyways.

Also, after some experimenting, it's very difficult to determine the correct number of goroutines that reduction operators should spawn. I could require that the caller specify the number of goroutines that should be spawned (as I did in the commits above before they were reverted), but that isn't very convenient for consumers of this library since it requires that they do their own benchmarking to find what's right for their use case.
